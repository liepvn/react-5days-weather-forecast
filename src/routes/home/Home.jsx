import React from 'react';
import { Col, Container, Row } from 'react-bootstrap';

import { WeatherForecast } from './components/WeatherForecast';
import { DEFAULT_LOCATION } from './home.constant';

/**
 * Home component
 *
 * @returns {JSX.Element}
 */
const Home = () => {
  return (
    <Container>
      <Row>
        <Col lg={{ span: 6, offset: 3 }} className='pt-4'>
          <WeatherForecast defaultLocation={DEFAULT_LOCATION} />
        </Col>
      </Row>
      <Row>
        <a
          href='https://github.com/liepvn'
          target='_blank'
          rel='noopener noreferrer'
          className='text-white text-center text-decoration-none mt-3'
        >
          <i className='bi bi-github'></i> <span>Fork it on Github</span>
        </a>
      </Row>
    </Container>
  );
};

export default Home;
