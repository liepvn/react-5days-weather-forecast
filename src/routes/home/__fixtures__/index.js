export * from './home-state.factory';