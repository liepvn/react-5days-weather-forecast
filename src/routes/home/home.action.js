import { createAction } from "core/helpers";

export const ActionTypes = {
  FETCH_WEATHER_FORECAST_REQUEST: 'FETCH_WEATHER_FORECAST_REQUEST',
  FETCH_WEATHER_FORECAST_SUCCESS: 'FETCH_WEATHER_FORECAST_SUCCESS',
  FETCH_WEATHER_FORECAST_FAILURE: 'FETCH_WEATHER_FORECAST_FAILURE',
  SEARCH_LOCATIONS_REQUEST: 'SEARCH_LOCATIONS_REQUEST',
  SEARCH_LOCATIONS_SUCCESS: 'SEARCH_LOCATIONS_SUCCESS',
  SEARCH_LOCATIONS_FAILURE: 'SEARCH_LOCATIONS_FAILURE',
}

export const fetchWeatherForecast = createAction(ActionTypes.FETCH_WEATHER_FORECAST_REQUEST, ({ woeid }) => ({ woeid }));
export const searchLocations = createAction(ActionTypes.SEARCH_LOCATIONS_REQUEST, (query) => ({ query }));