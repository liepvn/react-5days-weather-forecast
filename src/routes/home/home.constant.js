export const WEATHER_STATE_IMAGE_BASE_URL = 'https://www.metaweather.com/static/img/weather/png/64';

export const DEFAULT_LOCATION = Object.freeze({ woeid: 1252431, title: 'Ho Chi Minh City' });