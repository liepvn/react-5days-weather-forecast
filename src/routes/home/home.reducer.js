import { createReducer } from 'core/helpers';
import { take } from 'lodash';

import { ActionTypes } from './home.action';
import { WEATHER_STATE_IMAGE_BASE_URL } from './home.constant';

/**
 * Initial state
 */
export const homeState = {
  data: undefined,
  isSearching: false,
  isLoading: false,
  locations: [],
  forecast: []
};

/**
 * Reducer
 */
const homeReducer = createReducer(
  {
    [ActionTypes.FETCH_WEATHER_FORECAST_REQUEST]: (draft) => {
      draft.isLoading = true
    },
    [ActionTypes.FETCH_WEATHER_FORECAST_SUCCESS]: (draft, { payload }) => {
      const { consolidated_weather } = payload;

      draft.forecast = take(consolidated_weather, 5).map(
        ({ max_temp, min_temp, applicable_date, weather_state_abbr, humidity, the_temp }) => ({
          maxTemp: max_temp,
          minTemp: min_temp,
          date: applicable_date,
          weatherStateAbbr: weather_state_abbr,
          weatherStateImage: `${WEATHER_STATE_IMAGE_BASE_URL}/${weather_state_abbr}.png`,
          humidity,
          temp: the_temp
        })
      );
      draft.isLoading = false
    },
    [ActionTypes.FETCH_WEATHER_FORECAST_FAILURE]: (draft) => {
      draft.isLoading = false
    },
    [ActionTypes.SEARCH_LOCATIONS_REQUEST]: (draft) => {
      draft.isSearching = true
    },
    [ActionTypes.SEARCH_LOCATIONS_SUCCESS]: (draft, { payload }) => {
      draft.isSearching = false
      draft.locations = payload
    },
    [ActionTypes.SEARCH_LOCATIONS_FAILURE]: (draft) => {
      draft.isSearching = false
    },
  },
  homeState
);

const reducer = {
  home: homeReducer
}

export default reducer;
