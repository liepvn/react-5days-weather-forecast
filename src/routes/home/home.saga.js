import { all, call, put, takeLatest } from 'redux-saga/effects';
import weatherApi from 'shared/apis/weather.api';

import { ActionTypes } from './home.action';

/**
 * Handle fetching weather forecast data
 */
export function* fetchWeatherForecast({ payload }) {
  const { woeid } = payload || {};

  try {
    const { data } = yield call(weatherApi.fetchForecastByLocation, woeid);

    yield put({
      type: ActionTypes.FETCH_WEATHER_FORECAST_SUCCESS,
      payload: data
    });
  } catch (error) {
    yield put({
      type: ActionTypes.FETCH_WEATHER_FORECAST_FAILURE,
      payload: error
    });
  }
}

/**
 * Handle search locations
 */
export function* searchLocations({ payload: { query }}) {
  try {
    const { data } = yield call(weatherApi.searchLocations, query);

    yield put({
      type: ActionTypes.SEARCH_LOCATIONS_SUCCESS,
      payload: data
    });
  } catch (error) {
    yield put({
      type: ActionTypes.SEARCH_LOCATIONS_FAILURE,
      payload: error
    });
  }
}

export default function* root() {
  yield all([
    takeLatest(ActionTypes.FETCH_WEATHER_FORECAST_REQUEST, fetchWeatherForecast),
    takeLatest(ActionTypes.SEARCH_LOCATIONS_REQUEST, searchLocations)
  ]);
}
