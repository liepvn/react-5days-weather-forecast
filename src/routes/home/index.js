import Home from './Home';

import homeReducer from './home.reducer';
import homeSaga from './home.saga';

export { homeReducer, homeSaga };

export default Home;
