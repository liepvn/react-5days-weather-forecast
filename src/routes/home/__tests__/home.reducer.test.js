import { emptyAction } from "test/test-utils";

import hcmForecast from 'shared/apis/__fixtures__/weather-api-location-1252431.json';
import sanLocations from 'shared/apis/__fixtures__/weather-api-location-search-san.json';

import { ActionTypes, fetchWeatherForecast, searchLocations } from "../home.action";
import { default as reducer } from "../home.reducer";

describe('home.reducer', () => {
  let home = reducer.home(undefined, {});

  it('should return the initial state', () => {
    expect(reducer.home(home, emptyAction)).toMatchSnapshot();
  });

  it(`should handle ${ActionTypes.FETCH_WEATHER_FORECAST_REQUEST}`, () => {
    home = reducer.home(home, fetchWeatherForecast(1212));
    expect(home).toMatchSnapshot();
  });

  it(`should handle ${ActionTypes.FETCH_WEATHER_FORECAST_SUCCESS}`, () => {
    home = reducer.home(home, { type: ActionTypes.FETCH_WEATHER_FORECAST_SUCCESS, payload: hcmForecast });
    expect(home).toMatchSnapshot();
  });

  it(`should handle ${ActionTypes.SEARCH_LOCATIONS_REQUEST}`, () => {
    home = reducer.home(home, searchLocations('san'));
    expect(home).toMatchSnapshot();
  });

  it(`should handle ${ActionTypes.SEARCH_LOCATIONS_SUCCESS}`, () => {
    home = reducer.home(home, { type: ActionTypes.SEARCH_LOCATIONS_SUCCESS, payload: sanLocations });
    expect(home).toMatchSnapshot();
  });
});