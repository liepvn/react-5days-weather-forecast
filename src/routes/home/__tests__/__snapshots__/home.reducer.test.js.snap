// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home.reducer should handle FETCH_WEATHER_FORECAST_REQUEST 1`] = `
Object {
  "data": undefined,
  "forecast": Array [],
  "isLoading": true,
  "isSearching": false,
  "locations": Array [],
}
`;

exports[`home.reducer should handle FETCH_WEATHER_FORECAST_SUCCESS 1`] = `
Object {
  "data": undefined,
  "forecast": Array [
    Object {
      "date": "2021-08-15",
      "humidity": 73,
      "maxTemp": 32.06,
      "minTemp": 24.880000000000003,
      "temp": 30.515,
      "weatherStateAbbr": "hr",
      "weatherStateImage": "https://www.metaweather.com/static/img/weather/png/64/hr.png",
    },
    Object {
      "date": "2021-08-16",
      "humidity": 68,
      "maxTemp": 33.995000000000005,
      "minTemp": 24.725,
      "temp": 32.370000000000005,
      "weatherStateAbbr": "lr",
      "weatherStateImage": "https://www.metaweather.com/static/img/weather/png/64/lr.png",
    },
    Object {
      "date": "2021-08-17",
      "humidity": 70,
      "maxTemp": 33.825,
      "minTemp": 24.685,
      "temp": 31.630000000000003,
      "weatherStateAbbr": "hr",
      "weatherStateImage": "https://www.metaweather.com/static/img/weather/png/64/hr.png",
    },
    Object {
      "date": "2021-08-18",
      "humidity": 69,
      "maxTemp": 33.89,
      "minTemp": 25.35,
      "temp": 31.685000000000002,
      "weatherStateAbbr": "hr",
      "weatherStateImage": "https://www.metaweather.com/static/img/weather/png/64/hr.png",
    },
    Object {
      "date": "2021-08-19",
      "humidity": 70,
      "maxTemp": 33.835,
      "minTemp": 25.455,
      "temp": 31.515,
      "weatherStateAbbr": "hr",
      "weatherStateImage": "https://www.metaweather.com/static/img/weather/png/64/hr.png",
    },
  ],
  "isLoading": false,
  "isSearching": false,
  "locations": Array [],
}
`;

exports[`home.reducer should handle SEARCH_LOCATIONS_REQUEST 1`] = `
Object {
  "data": undefined,
  "forecast": Array [
    Object {
      "date": "2021-08-15",
      "humidity": 73,
      "maxTemp": 32.06,
      "minTemp": 24.880000000000003,
      "temp": 30.515,
      "weatherStateAbbr": "hr",
      "weatherStateImage": "https://www.metaweather.com/static/img/weather/png/64/hr.png",
    },
    Object {
      "date": "2021-08-16",
      "humidity": 68,
      "maxTemp": 33.995000000000005,
      "minTemp": 24.725,
      "temp": 32.370000000000005,
      "weatherStateAbbr": "lr",
      "weatherStateImage": "https://www.metaweather.com/static/img/weather/png/64/lr.png",
    },
    Object {
      "date": "2021-08-17",
      "humidity": 70,
      "maxTemp": 33.825,
      "minTemp": 24.685,
      "temp": 31.630000000000003,
      "weatherStateAbbr": "hr",
      "weatherStateImage": "https://www.metaweather.com/static/img/weather/png/64/hr.png",
    },
    Object {
      "date": "2021-08-18",
      "humidity": 69,
      "maxTemp": 33.89,
      "minTemp": 25.35,
      "temp": 31.685000000000002,
      "weatherStateAbbr": "hr",
      "weatherStateImage": "https://www.metaweather.com/static/img/weather/png/64/hr.png",
    },
    Object {
      "date": "2021-08-19",
      "humidity": 70,
      "maxTemp": 33.835,
      "minTemp": 25.455,
      "temp": 31.515,
      "weatherStateAbbr": "hr",
      "weatherStateImage": "https://www.metaweather.com/static/img/weather/png/64/hr.png",
    },
  ],
  "isLoading": false,
  "isSearching": true,
  "locations": Array [],
}
`;

exports[`home.reducer should handle SEARCH_LOCATIONS_SUCCESS 1`] = `
Object {
  "data": undefined,
  "forecast": Array [
    Object {
      "date": "2021-08-15",
      "humidity": 73,
      "maxTemp": 32.06,
      "minTemp": 24.880000000000003,
      "temp": 30.515,
      "weatherStateAbbr": "hr",
      "weatherStateImage": "https://www.metaweather.com/static/img/weather/png/64/hr.png",
    },
    Object {
      "date": "2021-08-16",
      "humidity": 68,
      "maxTemp": 33.995000000000005,
      "minTemp": 24.725,
      "temp": 32.370000000000005,
      "weatherStateAbbr": "lr",
      "weatherStateImage": "https://www.metaweather.com/static/img/weather/png/64/lr.png",
    },
    Object {
      "date": "2021-08-17",
      "humidity": 70,
      "maxTemp": 33.825,
      "minTemp": 24.685,
      "temp": 31.630000000000003,
      "weatherStateAbbr": "hr",
      "weatherStateImage": "https://www.metaweather.com/static/img/weather/png/64/hr.png",
    },
    Object {
      "date": "2021-08-18",
      "humidity": 69,
      "maxTemp": 33.89,
      "minTemp": 25.35,
      "temp": 31.685000000000002,
      "weatherStateAbbr": "hr",
      "weatherStateImage": "https://www.metaweather.com/static/img/weather/png/64/hr.png",
    },
    Object {
      "date": "2021-08-19",
      "humidity": 70,
      "maxTemp": 33.835,
      "minTemp": 25.455,
      "temp": 31.515,
      "weatherStateAbbr": "hr",
      "weatherStateImage": "https://www.metaweather.com/static/img/weather/png/64/hr.png",
    },
  ],
  "isLoading": false,
  "isSearching": false,
  "locations": Array [
    Object {
      "latt_long": "37.777119, -122.41964",
      "location_type": "City",
      "title": "San Francisco",
      "woeid": 2487956,
    },
    Object {
      "latt_long": "32.715691,-117.161720",
      "location_type": "City",
      "title": "San Diego",
      "woeid": 2487889,
    },
    Object {
      "latt_long": "37.338581,-121.885567",
      "location_type": "City",
      "title": "San Jose",
      "woeid": 2488042,
    },
    Object {
      "latt_long": "29.424580,-98.494614",
      "location_type": "City",
      "title": "San Antonio",
      "woeid": 2487796,
    },
    Object {
      "latt_long": "36.974018,-122.030952",
      "location_type": "City",
      "title": "Santa Cruz",
      "woeid": 2488853,
    },
    Object {
      "latt_long": "-33.463039,-70.647942",
      "location_type": "City",
      "title": "Santiago",
      "woeid": 349859,
    },
    Object {
      "latt_long": "36.406651,25.456530",
      "location_type": "City",
      "title": "Santorini",
      "woeid": 56558361,
    },
    Object {
      "latt_long": "43.461498,-3.810010",
      "location_type": "City",
      "title": "Santander",
      "woeid": 773964,
    },
    Object {
      "latt_long": "35.170429,128.999481",
      "location_type": "City",
      "title": "Busan",
      "woeid": 1132447,
    },
    Object {
      "latt_long": "28.46163,-16.267059",
      "location_type": "City",
      "title": "Santa Cruz de Tenerife",
      "woeid": 773692,
    },
    Object {
      "latt_long": "35.666431,-105.972572",
      "location_type": "City",
      "title": "Santa Fe",
      "woeid": 2488867,
    },
  ],
}
`;

exports[`home.reducer should return the initial state 1`] = `
Object {
  "data": undefined,
  "forecast": Array [],
  "isLoading": false,
  "isSearching": false,
  "locations": Array [],
}
`;
